File: python/promethean/path_finder.py
from typing import List, Optional

from promethean.options import PathFinderTile, PathFinderOptions
from promethean.point import Point


class PathFinderNode:
    def __init__(self, position: Point, g: int, h: int, parent_node_position: Point):
        self._position: Point = position
        self._g: int = g
        self._h: int = h
        self._parent_node_position: Point = parent_node_position
        self._f: int = -1  # non-defined value
        self._closed: Optional[bool] = None  # make True when we remove the node from closed list, make False when we add the node to the open list

    def reset(self):
        self._g = 0
        self._h = 0
        self._parent_node_position = Point()
        self._f = -1
        self._closed = None

    def update(self, g: int, h: int, parent: Point):
        if self._f == -1 or self._f > g + h:
            self._g = g
            self._h = h
            self._f = g + h
            self._parent_node_position = parent

    def close(self):
        self._closed = True

    def open(self):
        self._closed = False

    def is_closed(self) -> bool:
        if self._closed is None:
            return False
        else:
            return self._closed

    def is_undefined(self) -> bool:
        if self._closed is None:
            return True
        else:
            return False

    def __repr__(self) -> str:
        return "<" + str(self._position) + ":" + str(self._g) + ":" + str(self._h) + ":" + str(self._parent_node_position) + ":" + str(self._closed) + ">"


def distance(start: Point, end: Point) -> int:
    return abs(start._x - end._x) + abs(start._y - end._y)


class PathFinderGraph:
    def __init__(self, height: int, width: int):
        self._height: int = height
        self._width: int = width
        self._internal_grid: List[List[PathFinderNode]] = [[PathFinderNode(Point(i, j), 0, 0, Point()) for j in range(width)]for i in range(height)]
        self._open: List[PathFinderNode] = []

    def reset(self):
        for i in range(self._height):
            for j in range(self._width):
                self._internal_grid[i][j].reset()
        self._open.clear()

    def open_node(self, position: Point, g: int, h: int, parent: Point):
        node: PathFinderNode = self._internal_grid[position._x][position._y]
        node.update(g, h, parent)
        node.open()
        self._open.append(node)

    def has_open_nodes(self) -> bool:
        return len(self._open) > 0

    def get_open_node_with_smallest_f(self) -> Optional[PathFinderNode]:
        if len(self._open) > 0:
            to_return_i: int = 0
            to_return_f: int = self._open[0]._f
            for i, node in enumerate(self._open):
                if node._f < to_return_f:
                    to_return_i = i
                    to_return_f = node._f
            # set the node closed
            return_node: PathFinderNode = self._open.pop(to_return_i)
            return_node.close()
            return return_node
        else:
            return None

    def add_node(self, x: int, y: int, g: int, target: Point, parent: Point):
        node: PathFinderNode = self._internal_grid[x][y]
        h: int = distance(node._position, target)
        if not node.is_closed():
            node.update(g, h, parent)
            # add to the open list
            if node.is_undefined():
                node.open()
                self._open.append(node)


def order_closed_nodes_as_array(graph: PathFinderGraph, end_node: PathFinderNode) -> List[Point]:
    current_node: PathFinderNode = end_node
    to_return: List[Point] = [Point() for i in range(end_node._g + 1)]
    for i in range(end_node._g + 1):
        to_return[len(to_return) - i - 1] = current_node._position
        current_node = graph._internal_grid[current_node._parent_node_position._x][current_node._parent_node_position._y]
    return to_return


class PathFinder:
    def __init__(self, world_grid: List[List[PathFinderTile]], options: PathFinderOptions):
        self._world_grid = world_grid
        self._options = options
        self._height = len(world_grid)
        self._width = len(world_grid[0])
        self._graph: PathFinderGraph = PathFinderGraph(self._height, self._width)

    def block_point(self, point: Point):
        self._world_grid[point._x][point._y] = PathFinderTile.Blocked

    def find_path(self, start: Point, end: Point) -> List[Point]:
        nodes_visited: int = 0
        self._graph.reset()
        self._graph.open_node(start, 0, distance(start, end), start)

        while self._graph.has_open_nodes():
            q: Optional[PathFinderNode] = self._graph.get_open_node_with_smallest_f()
            if q is not None:
                if q._position == end:
                    return order_closed_nodes_as_array(self._graph, q)

                if nodes_visited > self._options.search_limit:
                    return []

                # for each node we should select 4 adjacent walkable points
                # and add it to the open list, if the point is not visited
                x: int = q._position._x
                y: int = q._position._y
                g = q._g + 1

                if self._world_grid[x][y - 1] == PathFinderTile.Pathable:
                    self._graph.add_node(x, y - 1, g, end, q._position)
                if self._world_grid[x][y + 1] == PathFinderTile.Pathable:
                    self._graph.add_node(x, y + 1, g, end, q._position)
                if self._world_grid[x - 1][y] == PathFinderTile.Pathable:
                    self._graph.add_node(x - 1, y, g, end, q._position)
                if self._world_grid[x + 1][y] == PathFinderTile.Pathable:
                    self._graph.add_node(x + 1, y, g, end, q._position)
            else:
                break

            nodes_visited += 1

        return []

    def __repr__(self) -> str:
        return "\n".join([" ".join([str(int(self._world_grid[i][j])) for j in range(self._width)]) for i in range(self._height)])

File: rust/promethean/src/path_finder.rs
use crate::level::point::Point;
use crate::level::options::PathFinderTile;
use std::fmt;

#[derive(Debug)]
pub struct PathFinderNode {
    position: Point,
    g: usize,
    h: usize,
    parent_node_position: Point,
    f: usize,
    closed: u8
}

impl PathFinderNode{
    pub fn new(position: Point, g: usize, h: usize, parent_node_position: Point) -> PathFinderNode {
        return PathFinderNode{ position: position,
                               g: g,
                               h: h,
                               parent_node_position: parent_node_position,
                               f: 0,
                               closed: 0};
    }

    pub fn clone(&self) -> PathFinderNode {
        return PathFinderNode { position: self.position.clone(),
                                g: self.g,
                                h: self.h,
                                parent_node_position: self.parent_node_position.clone(),
                                f: self.f,
                                closed: self.closed};
    }

    pub fn f(&self) -> usize{
        return self.f;
    }

    pub fn g(&self) -> usize{
        return self.g;
    }

    pub fn position(&self) -> &Point{
        return &self.position;
    }

    pub fn parent_position(&self) -> &Point{
        return &self.parent_node_position;
    }

    pub fn reset(&mut self) {
        self.g = 0;
        self.h = 0;
        self.parent_node_position.set(0, 0);
        self.f = 0;
        self.closed = 0;
    }

    pub fn update(&mut self, g: usize, h: usize, parent: Point) {
        if self.f == 0 || self.f > g + h{
            self.g = g;
            self.h = h;
            self.f = g + h;
            self.parent_node_position = parent;
        }
    }

    pub fn close(&mut self) {
        self.closed = 2;
    }

    pub fn open(&mut self) {
        self.closed = 1;
    }

    pub fn is_closed(&self) -> bool {
        if self.closed != 0{
            return self.closed == 2
        }
        else {
            return false;
        }
    }

    pub fn is_undefined(&self) -> bool {
        if self.closed != 0 {
            return false;
        }
        else {
            return true;
        }
    }
}

impl fmt::Display for PathFinderNode{
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        return write!(f, "<{}:{}:{}:{}:{}>", self.position, self.g, self.h, self.parent_node_position, self.closed);
    }
}

pub fn distance(start: &Point, end: &Point) -> usize{
    let sx = start.x();
    let sy = start.y();
    let ex = end.x();
    let ey = end.y();
    let x: usize = if sx > ex { (sx - ex) as usize } else { (ex - sx) as usize };
    let y: usize = if sy > ey { (sy - ey) as usize } else { (ey - sy) as usize };
    return x + y;
}

#[derive(Debug)]
pub struct PathFinderGraph {
    height: usize,
    width: usize,
    internal_grid: Vec<Vec<PathFinderNode>>,
    open: Vec<Point>
}

impl PathFinderGraph {
    pub fn new(height: usize, width: usize) -> PathFinderGraph {
        let mut internal_grid = Vec::with_capacity(height);
        for x in 0..height {
            let mut x_array = Vec::with_capacity(width);
            for y in 0..width {
                x_array.push(PathFinderNode::new(Point::new(x as i32, y as i32), 0, 0, Point::new(0, 0)));
            }
        internal_grid.push(x_array);
        }

        let open = Vec::with_capacity(width);

        return PathFinderGraph{ height, width, internal_grid, open};
    }

    pub fn get_node_from_grid(&self, x: usize, y: usize) -> &PathFinderNode {
        return &self.internal_grid[x][y];
    }

    pub fn reset(&mut self) {
        for x in 0.. self.height {
            for y in 0.. self.width {
                self.internal_grid[x][y].reset();
            }
        }
        self.open.clear();
    }

    pub fn open_node(&mut self, position: &Point, g: usize, h: usize, parent: &Point) {
        let node: &mut PathFinderNode = &mut self.internal_grid[position.x() as usize][position.y() as usize];
        node.update(g, h, Point::new(parent.x(), parent.y()));
        node.open();

        self.open.push(Point::new(node.position.x(), node.position.y()));
    }

    pub fn has_open_nodes(&self) -> bool {
        return self.open.len() > 0;
    }

    pub fn get_open_node_with_smallest_f(&mut self) -> PathFinderNode {
        let zero_point: &Point = &self.open[0];
        let mut to_return_i: usize = 0;
        let mut to_return_f: usize = self.internal_grid[zero_point.x() as usize][zero_point.y() as usize].f();

        for i in 0..self.open.len() {
            let i_point: &Point = &self.open[i];
            let node: &PathFinderNode = &self.internal_grid[i_point.x() as usize][i_point.y() as usize];
            let node_f: usize = node.f();

            if node_f < to_return_f {
                to_return_i = i;
                to_return_f = node_f;
            }
        }

        let zero_point = &self.open[to_return_i];
        let mut return_node: PathFinderNode = self.internal_grid[zero_point.x() as usize][zero_point.y() as usize].clone();
        self.open.remove(to_return_i);

        return_node.close();

        return return_node;
    }

    pub fn add_node(&mut self, x: usize, y: usize, g: usize, target: &Point, parent: &Point) {
        let node: &mut PathFinderNode = &mut self.internal_grid[x][y];
        let h = distance(node.position(), target);
        if !node.is_closed() {
            node.update(g, h, Point::new(parent.x(), parent.y()));
            if node.is_undefined() {
                node.open();
                self.open.push(Point::new(node.position.x(), node.position.y()));
            }
        }
    }
}

pub fn order_closed_nodes_as_array(graph: &PathFinderGraph, end_node: &PathFinderNode) -> Vec<Point> {
    let mut current_node: &PathFinderNode = end_node;
    let length = end_node.g() + 1;
    let mut to_return = Vec::with_capacity(length);

    for _ in 0..length {
        to_return.push(current_node.position().clone());
        let pnp: &Point = current_node.parent_position();
        current_node = graph.get_node_from_grid(pnp.x() as usize, pnp.y() as usize);
    }
    to_return.reverse();
    return to_return;
}

#[derive(Debug)]
pub struct PathFinder {
    world_grid: Vec<Vec<PathFinderTile>>,
    search_limit: usize,
    height: usize,
    width: usize,
    graph: PathFinderGraph
}

impl PathFinder {
    pub fn new(world_grid: Vec<Vec<PathFinderTile>>, search_limit: usize) -> PathFinder {
        let height = world_grid.len();
        let width = world_grid[0].len();
        return PathFinder { world_grid, search_limit, height, width, graph: PathFinderGraph::new(height, width)};
    }

    pub fn block_point(&mut self, point: Point) {
        self.world_grid[point.x() as usize][point.y() as usize] = PathFinderTile::Blocked;
    }

    pub fn find_path(&mut self, start: &Point, end: &Point) -> Vec<Point> {
        let mut nodes_visited: usize = 0;
        self.graph.reset();
        self.graph.open_node(start, 0, distance(start, end), start);

        while self.graph.has_open_nodes() {
            let q: PathFinderNode = self.graph.get_open_node_with_smallest_f();
            if q.position().equal(end) {
                return order_closed_nodes_as_array(&self.graph, &q);
            }

            if nodes_visited > self.search_limit {
                return Vec::new();
            }

            let x: usize = q.position.x() as usize;
            let y: usize = q.position.y() as usize;
            let g: usize = q.g() + 1;

            match self.world_grid[x][y - 1] {
                PathFinderTile::Pathable => self.graph.add_node(x, y - 1, g, end, q.position()),
                PathFinderTile::Blocked => {}
            }
            match self.world_grid[x][y + 1] {
                PathFinderTile::Pathable => self.graph.add_node(x, y + 1, g, end, q.position()),
                PathFinderTile::Blocked => {}
            }
            match self.world_grid[x - 1][y] {
                PathFinderTile::Pathable => self.graph.add_node(x - 1, y, g, end, q.position()),
                PathFinderTile::Blocked => {}
            }
            match self.world_grid[x + 1][y] {
                PathFinderTile::Pathable => self.graph.add_node(x + 1, y, g, end, q.position()),
                PathFinderTile::Blocked => {}
            }

            nodes_visited += 1;
        }
        return Vec::new();
    }
}

impl fmt::Display for PathFinder{
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut to_return = String::new();
        for x in 0..self.height {
            let mut x_str = String::new();
            for y in 0..self.width {
                x_str.push_str(match self.world_grid[x][y] {PathFinderTile::Blocked => "■ ", PathFinderTile::Pathable =>  "□ " });
            }
            to_return.push_str((x_str + "\n").as_str());
        }

        return write!(f, "{}", to_return);
    }
}